File: VB/reporting-provide-custom-report-gallery-storage-in-report-designer/XpoReportGalleryStorage/XpoReportGalleryStorageSimple.vb
Imports System.Linq
Imports DevExpress.Xpo
Imports DevExpress.XtraReports.ReportGallery
Imports System.Runtime.InteropServices

Namespace XpoReportGalleryStorage

    Public Class XpoReportGalleryStorageSimple
        Inherits XpoReportGalleryStorageBase

        ' Enable synchronization with a remote storage.
        Public Overrides ReadOnly Property EnableSynchronization As Boolean
            Get
                Return True
            End Get
        End Property

        Public Overrides Sub SaveGallery(ByVal gallery As Gallery)
            Using unitOfWork As UnitOfWork = New UnitOfWork(DataLayer)
                unitOfWork.BeginTransaction()
                Dim deletedItems = unitOfWork.Query(Of ReportGalleryTable)().[Select](Function(a) a)
                For Each deletes In deletedItems
                    deletes.Delete()
                Next

                unitOfWork.CommitTransaction()
                unitOfWork.BeginTransaction()
                For Each newItem In gallery.Items
                    MapFromGalleryItem(newItem, unitOfWork)
                Next

                unitOfWork.CommitTransaction()
            End Using
        End Sub

        Public Overrides Function TryLoadGallery(<Out> ByRef gallery As Gallery) As Boolean
            Using unitOfWork As UnitOfWork = New UnitOfWork(DataLayer)
                gallery = New Gallery()
                Dim remoteItems = unitOfWork.Query(Of ReportGalleryTable)().[Select](Function(a) MapFromTableRow(a)).ToList()
                If Not remoteItems.Any() Then Return False
                gallery.Items.AddRange(remoteItems)
                Return True
            End Using
        End Function
    End Class
End Namespace

File: CS/XpoReportGalleryStorage/XpoReportGalleryStorageSimple.cs
using System.Linq;
using DevExpress.Xpo;
using DevExpress.XtraReports.ReportGallery;

namespace XpoReportGalleryStorage {
    public class XpoReportGalleryStorageSimple : XpoReportGalleryStorageBase {

        // Enable synchronization with a remote storage.
        public override bool EnableSynchronization => true;

        public override void SaveGallery(Gallery gallery) {
            using(UnitOfWork unitOfWork = new UnitOfWork(DataLayer)) {
                unitOfWork.BeginTransaction();
                var deletedItems = unitOfWork.Query<ReportGalleryTable>().Select(a => a);
                foreach(var deletes in deletedItems) {
                    deletes.Delete();
                }
                unitOfWork.CommitTransaction();
                unitOfWork.BeginTransaction();
                foreach(var newItem in gallery.Items)
                    MapFromGalleryItem(newItem, unitOfWork);
                unitOfWork.CommitTransaction();
            }
        }

        public override bool TryLoadGallery(out Gallery gallery) {
            using(UnitOfWork unitOfWork = new UnitOfWork(DataLayer)) {
                gallery = new Gallery();
                var remoteItems = unitOfWork.Query<ReportGalleryTable>().Select(a => MapFromTableRow(a)).ToList();
                if(!remoteItems.Any())
                    return false;
                gallery.Items.AddRange(remoteItems);
                return true;
            }
        }
    }
}
